### PR TITLE
17 attack

### PR DESCRIPTION
**Pull request by philswatton**
This PR implements #17 and #18. To do this, it:

- Adds functions for attacks and transfer attacks in a new attack module
- Adds a function to load models and restore runs from wandb
- Adds a generation script (similar to those used for similarity metrics and model training)
- Updates other files to facilitate this
- Adds foolbox dependency
- Adds tests

EDIT: some stuff that's added doesn't include tests, these are:

- Model loading (just a wrapper for wandb)
- Attack generation(partially a wrapper with some extra logic, needs a working model to run)
- Transfering the attacks (mostly a wrapper - metrics calculation is handled in other functions which are tested)

### CODE REVIEW

File: src/modsim2/attack/attack.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        raise NotImplementedError("This attack has not been implemented")
+
+    # Put model into foolbox format
+    fmodel = fb.PyTorchModel(model, bounds=(0, 1))
+
+    # Generate attack images
+    attack = getattr(fb.attacks, attack_fn_name)(**kwargs)
+    _, clipped_advs, success = attack(fmodel, images, labels, epsilons=epsilons)
+
+    # Apply image selection based on attack choice
```
[the remark this answers is not in the document]
Comment (lannelin) [reply src/modsim2/attack/attack.py:42]: we should also document this behaviour clearly in README (or point to an issue from README or similar)

File: scripts/attack.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        "dist_B": transfer_metrics_BB_to_A,
+    }
+    run_A.log({"A_to_B_metrics": A_to_B_metrics}, commit=True)
+    run_B.log({"B_to_A_metrics": B_to_A_metrics}, commit=True)
+
+    # Close once finished
+    run_A.finish()
+    run_B.finish()
+
+    return A_to_B_metrics, B_to_A_metrics
```
Comment (lannelin) [inline scripts/attack.py:131]: don't return here as we don't intend to call `main` to get a value back

File: tests/test_transfer_metrics.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        compute_mean_loss_rate(
+            labels=labels,
+            base_loss=base_loss,
+            advs_softmax=base_softmax,
+            loss_function=f,
+        )
+        == 0
+    )
+
+    # Brittle but useful test
```
Comment (lannelin) [inline tests/test_transfer_metrics.py:44]: mysterious comment, why is this brittle?

File: tests/test_data.py [hunk 4]
```diff
@@ -112,6 +112,29 @@ def test_cifar_stratification():
...
+    dmpair = DMPair(drop_percent_A=drop_A, drop_percent_B=drop_B)
+
+    # Sample from test
+    num_image = 20
+    _, test_labels_A, _, test_labels_B = dmpair.sample_from_test_pairs(num_image)
+
+    # Get full labels
+    full_test_labels = [image[1] for image in dmpair.cifar.dataset_test]
+
+    # Use the fact that sampling 20/100 is equiv to dropping 80
```
Comment (lannelin) [inline tests/test_data.py:127]: be a bit more explicit here. comment that we can reuse the dm_stratification test as this is the same case as dropping 80

**Review by lannelin — COMMENTED** [review]
nice. Happy to discuss any of the comments, especially the wandb id one!

haven't had the chance to run this fully with real data, would be keen to see that done before we merge.

"adversial"->"adversarial" throughout

File: scripts/generate_attack_scripts.py
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/generate_attack_scripts.py:1]: Per our discussion, I'm going to keep all this information in the scripts/README.md alone so there is a single source of truth on the scripts

File: src/modsim2/attack/attack.py
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/attack.py:1]: have renamed to adversarial_images

File: src/modsim2/attack/attack.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    Returns: A dictionary of 4 sets of adverisal images
+    """
+    # Get num of images
+    num_attack_images = len(images_A)
+
+    # Make dict of adversial images
+    # 4 elements, w/ keys like model_A_dist_A
+    # Each element is a torch.tensor containing adversial images
+    adversial_images_dict = {}
+    for model in [(model_A, "model_A"), (model_B, "model_B")]:
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/attack.py:96]: done

File: src/modsim2/attack/attack.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        for image_set in [
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/attack.py:97]: done

File: src/modsim2/attack/transfer.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    percentage of images which in their base form are correctly identified by the
+    network but which in their adversial form are not.
+
+    Args:
+        labels: The correct labels
+        base_correct: List denoting which base images the network correctly
+                      predicts
+        advs_preds: List of predictions for the adversial images
+    """
+    advs_correct = (labels == advs_preds)[base_correct]
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/transfer.py:24]: We use all images in the other metric - possibly worth discussing whether we'd want that though

[the remark this answers is not in the document]
Comment (lannelin) [reply src/modsim2/attack/transfer.py:24]: ah no that makes sense, sorry!

File: src/modsim2/attack/attack.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    images_A: torch.tensor,
+    labels_A: torch.tensor,
+    images_B: torch.tensor,
+    labels_B: torch.tensor,
+    attack_fn_name: str,
+    **kwargs,
+) -> dict[torch.tensor]:
+    """
+    This function loops over a 2*2 combination of models (A and B) and image/label
+    pairs (A and B) to produce 4 sets of adversial images.
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/attack.py:68]: done

File: tests/test_data.py (hunk 4 above)
Comment (philswatton) [reply tests/test_data.py:127]: done

File: src/modsim2/attack/attack.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/attack.py:42]: I've extracted the function and added a test for it, will add to README

File: scripts/attack.py (hunk 2 above)
Comment (philswatton) [reply scripts/attack.py:131]: done

File: configs/trainer.yaml [hunk 9]
```diff
@@ -10,3 +10,4 @@ wandb:
   entity: turing-arc
   project: ms2
   mode: online
+  id_postfix: _test_run #this postfix MUST be changed on each run
```
[the remark this answers is not in the document]
Comment (philswatton) [reply configs/trainer.yaml:13]: Following our discussion yesterday, I've added the utility functions to ensure that a given run name cannot exist prior to a model being trained, and must be unique prior to transfer attacks being done. This will enforce uniqueness on the user. I'll make sure to add to README

File: scripts/README.md [hunk 10]
```diff
@@ -46,18 +48,43 @@ sbatch trainer_scripts/drop-only_0_0_trainer.sh
 
 to train models for that particular DMPair. Note that logging is performed with wandb and so you will need this set up on your machine in order to proceed.
 
-### Requirements
+## Transfer Attacks
+
+Call:
+
+```bash
+python scripts/generate_attack_scripts.py --experiment_group "drop-only"
```
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/README.md:70]: done

File: scripts/README.md
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/README.md:1]: I've fleshed out the README - let me know if it needs more!

File: src/modsim2/attack/transfer.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        base_correct: List denoting which base images the network correctly
+                      predicts
+        advs_preds: List of predictions for the adversial images
+    """
+    advs_loss = loss_function(advs_softmax, labels)
+    mean_loss_rate = advs_loss - base_loss
+    return mean_loss_rate
+
+
+def compute_transfer_attack(
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/transfer.py:50]: I've added this

File: src/modsim2/attack/transfer.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        labels: A list of correct labels corresponding to each image
+        advs_images: A torch.tensor of adversial images, which corresponds to the base
+                     images
+        attack_names: Output strings for the attack names
+        loss_function: Loss function to use in computing mean_loss_rate
+
+    Returns: a dictionary of attack success metrics
+    """
+    # Generate base image and attack image softmax and predictions
+    base_softmax = model.forward(images)
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/transfer.py:85]: Now using `trainer.predict()` for transfer attacks

File: src/modsim2/attack/transfer.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    base_preds = torch.max(base_softmax, dim=1)[1]
+
+    # These are recycled for each attack, worth computing now
+    base_correct = labels == base_preds
+    base_loss = loss_function(base_softmax, labels)
+
+    # Generate adversarial softmax and predictions
+    advs_softmax = [model.forward(images) for images in advs_images]
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/transfer.py:93]: Now using `trainer.predict()` for transfer attacks

File: src/modsim2/attack/attack.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/attack/attack.py]: Done

File: scripts/generate_attack_scripts.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        for dataset_index in range(NUM_PAIRS)
+    ]
+    script_names = [
+        scripts_path + "/" + experiment_name + "_attack.sh"
+        for experiment_name in experiment_names
+    ]
+
+    # For each combination, write bash script with params
+    for index, combo in enumerate(combinations):
+        python_call = f"python scripts/attack.py {combo}"
```
[the remark this answers is not in the document]
Comment (philswatton) [reply scripts/generate_attack_scripts.py:54]: Done

File: configs/attack.yaml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+num_attack_images: 16
```
[the remark this answers is not in the document]
Comment (philswatton) [reply configs/attack.yaml]: Done

**Review by lannelin — APPROVED** [review]
nice work, very minor comments and happy to be merged directly after.

File: src/modsim2/model/utils.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    project_name: str,
+) -> bool:
+    """
+    A convinience function for checking if a given wandb run name already exists
+    on wandb. Returns True or False
+
+    Args:
+        run_name: the name of the run to check for existence
+    """
+    api = wandb.Api()
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/model/utils.py]: Done!

File: tests/test_adversarial_images.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+    # a tensor of successes, where all epsilons above 0 are successful
+    success = torch.zeros((num_epsilon, num_image), dtype=torch.bool)
+    success[[0, 1, 3], :] = True
+
+    # this means the best selection function should *always* select the
+    # second image in the list (as this corresponds to the lowest epsilon 0.3
+    # that is still success)
+    best_images = select_best_attack(images=images, success=success, epsilons=epsilons)
+    assert torch.all(images[1] == best_images).item()
```
[the remark this answers is not in the document]
Comment (philswatton) [reply tests/test_adversarial_images.py]: And done!
